sync-core/src/main/java/com/cloudant/sync/query/QueryResult.java

```diff
@@ -55,109 +55,129 @@ public QueryResult(List<String> docIds,
+                List<BasicDocumentRevision> docs = datastore.getDocumentsWithIds(batch);
+                for (DocumentRevision rev : docs) {
+                    DocumentRevision innerRev;
+                    innerRev = rev;  // Allows us to replace later if projecting
+
+                    // Apply post-hoc matcher
+                    if (matcher != null && !matcher.matches(innerRev)) {
+                        continue;
+                    }
+
+                    // Apply skip (skip == 0 means disable)
+                    if (skip > 0 && nSkipped < skip) {
+                        nSkipped = nSkipped + 1;
+                        continue;
+                    }
+
+                    // TODO - Add projection logic
+
+                    docList.add(innerRev);
+
+                    // Apply limit (limit == 0 means disable)
+                    nReturned = nReturned + 1;
+                    if (limit > 0 && nReturned >= limit) {
+                        stop = true;
+                        break;
+                    }
 
-            /**
-             * Partition a list of document IDs into batches of batchSize.
-             *
-             * Return a mutable list of consecutive sublists.
-             * Same as Guava's "Lists.partition" except the result list is mutable.
-             * It is needed because this iterator removes sublist from the partitions
-             * as it goes.
-             *
-             * @see <a href="http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/
-             * common/collect/Lists.html#partition(java.util.List, int)">Lists.partition</a>
-             */
-            private List<List<String>> partition(List<String> documentIds, int batchSize) {
-                List<List<String>> partitions = Lists.partition(documentIds, batchSize);
-                List<List<String>> res = new LinkedList<List<String>>();
-                for(List<String> p : partitions) {
-                    res.add(p);
                 }
-                return res;
-            }
 
-            /**
-             * Load the next partition of DocumentRevision objects for the
-             * iterator.
-             *
-             * @param ids the IDs of the revisions to load.
-             * @return an iterator over the DocumentRevision objects for `ids`.
-             */
-            private Iterator<BasicDocumentRevision> nextSubIterator(List<String> ids) {
-                HashMap<String, BasicDocumentRevision> map = new HashMap<String, BasicDocumentRevision>();
-                for(BasicDocumentRevision revision : datastore.getDocumentsWithIds(ids)) {
-                    map.put(revision.getId(), revision);
+                range.location = range.location + range.length;
+
+                if (stop) {
+                    // Limit reached.  Move range location to the end.
+                    range.location = originalDocIds.size();
```

Comment:
I'd not use this technique of pretending the end of the list has been reached when really you've reached the limit, instead I'd make `stop` class level and make the hasNext check include it: 

```java
if (stop) {
    return false;
} else {
    return (documentBlock.hasNext() || range.location < originalDocIds.size());
}
```

Perhaps renamed `stop` to `limitReached`.